File: solutions/haskell/yacht/3/src/Yacht.hs
module Yacht (yacht, Category(..)) where
import Data.List (find, nub, sort)

data Category = Ones
              | Twos
              | Threes
              | Fours
              | Fives
              | Sixes
              | FullHouse
              | FourOfAKind
              | LittleStraight
              | BigStraight
              | Choice
              | Yacht deriving (Enum)

yacht :: Category -> [Int] -> Int
yacht Yacht dice | length (nub dice) == 1 = 50
yacht FullHouse dice | sort (map snd (frequencies dice)) == [2,3] = sum dice
yacht FourOfAKind dice
  | sort (map snd (frequencies dice)) == [1,4] =
      case find (\x -> snd x == 4) (frequencies dice) of
        Just v -> 4 * fst v
        Nothing -> 0 
  | sort (map snd (frequencies dice)) == [5] =
    case find (\x -> snd x == 5) (frequencies dice) of
        Just v -> 4 * fst v
        Nothing -> 0
yacht LittleStraight dice | sort dice == [1,2,3,4,5] = 30
yacht BigStraight dice | sort dice == [2,3,4,5,6] = 30
yacht category dice | fromEnum category <= 5 = sum (filter (== (fromEnum category + 1)) dice)
yacht Choice dice = sum dice
yacht _ _ = 0

frequencies :: [Int] -> [(Int, Int)]
frequencies [] = []
frequencies (x:xs) = (x, length (filter (== x) (x:xs))) : frequencies (filter (/= x) xs)

File: solutions/erlang/yacht/1/src/yacht.erl
-module(yacht).
-import(lists, [filter/2, map/2, sort/1, sum/1]).
-export([score/2, frequencies/1]).

score(D, yacht) ->
  F = frequencies(D),
  case length(F) of
    1 -> 50;
    _ -> 0
  end;
score(D, ones)   -> sumElemsEq(1, D);
score(D, twos)   -> sumElemsEq(2, D);
score(D, threes) -> sumElemsEq(3, D);
score(D, fours)  -> sumElemsEq(4, D);
score(D, fives)  -> sumElemsEq(5, D);
score(D, sixes)  -> sumElemsEq(6, D);
score(D, four_of_a_kind) ->
  F = frequencies(D),
  case (filter(fun ({_,FR}) -> (FR == 4) or (FR == 5) end, F)) of
    [{N,4}] -> 4*N;
    [{N,5}] -> 4*N;
    _ -> 0
  end;
score(D, little_straight) ->
  case sort(D) of
    [1,2,3,4,5] -> 30;
    _ -> 0
  end;
score(D, big_straight) ->
  case sort(D) of
    [2,3,4,5,6] -> 30;
    _ -> 0
  end;
score(D, full_house) ->
  F = frequencies(D),
  case sort(map(fun snd/1, F)) of
    [2,3] -> sum(D);
    _ -> 0
  end;
score(Dice, choice) -> sum(Dice).

frequencies([]) -> [];
frequencies([H|T]) ->
  CountT = length(filter(fun(E) -> E =:= H end, T)),
  NewT = filter(fun(E) -> E =/= H end, T),
  [{H, 1 + CountT} | frequencies(NewT)].

sumElemsEq(N, XS) -> sum(filter(fun(E) -> E =:= N end, XS)).

snd({_,S}) -> S.
